core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,146 @@
+package ca.jrvs.apps.grep;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+import org.apache.log4j.BasicConfigurator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JavaGrepImp implements JavaGrep {
+
+  final Logger logger = LoggerFactory.getLogger(JavaGrep.class);
+
+  private String regex;
+  private String rootPath;
+  private String outFile;
+
+  @Override
+  public void process() throws IOException {
+    try {
+      List<String> matchedLines = new ArrayList<>();
+
+      for (File file : listFiles(getRootPath())) {
+        for (String line : readLines(file)) {
+          if (containsPattern(line)) {
+            matchedLines.add(line);
+          }
+        }
+      }
+      writeToFile(matchedLines);
+    } catch (IOException ex) {
+      logger.error("USAGE: JavaGrep regex rootPath outFile");
+    }
+  }
+
+  @Override
+  public List<File> listFiles(String rootDir) throws IOException {
+    List<File> files = new ArrayList<>();
+    File[] dir = new File(rootDir).listFiles();
+
+    for (File file : dir) {
+      if (file.isDirectory()) {
+        files.addAll(listFiles(file.getAbsolutePath()));
+      } else if (file.isFile()) {
+        files.add(file);
+      }
+    }
+    return files;
+  }
+
+  @Override
+  public List<String> readLines(File inputFile) throws IOException {

```

Comment:
try not to return multiple times for better readablity 

String line = null;

try{
  line = blah
}

return lines